Source/IronSourceAdapter.swift

```diff
@@ -50,16 +50,21 @@ final class IronSourceAdapter: PartnerAdapter {
             completion(error)
             return
         }
-        // Initialize IronSource
-        IronSource.initISDemandOnly(appKey, adUnits: configuration.lineItems ?? [])
-        
-        // IronSource provides one single delegate for all ads of the same type.
-        // IronSourceAdapterRouter implements these delegate protocols and forwards calls to the corresponding partner ad instances.
+
+        // Initialize IronSource. Must be performed on the main queue.
+        dispatchPrecondition(condition: .notOnQueue(.main)) // If on the main queue, there will be deadlock. Make sure we are not on the main queue.
+        DispatchQueue.main.sync {
```

Comment:
Can we just dispatch async on the main queue, so we don't cause a deadlock if in the future the SDK calls this method from the main queue?

Reply:
Then you have to make a choice:

1. make the completion callback on the main queue (not the original queue)
2. make the completion call back outside the dispatch which could occur out of order.

Reply:
Then you have to make a choice:

1. make the completion callback on the main queue (not the original queue)
2. make the completion callback outside the dispatch which could occur out of order.

I swear I made the right call here.

Also consider:

1. Why would you change the future SDK calls do all that work on the main queue? I find that extremely unlikely.
2. This assertion will help find that issue...it's the point of the assertion.

Reply:
I thinking making the completion callback on the main queue is fine.
Doing it async has the added benefit of not blocking the current queue.
Even if it's unlikely, I think it's better to not make and enforce this assumption here. I don't see anything necessarily wrong with us deciding at some point we want to make all these calls on the main queue for all adapters. Having this adapter police the SDK behavior doesn't seem necessary to me.

If you still feel strongly please ping the rest of the team and we can get a 3rd voice.

Reply:
Yeah, I want more opinions.  Only because I would not have imagined being challenged with this minor change. To me, what I did here I thought was perfectly acceptable -- I was thinking to myself there is no way anybody would accept the callback not being on the original queue.

Reply:
@kendallrogers to chime in

Reply:
I've given a lot of thought about this unfortunately overnight. I'm firm here. The argument that the SDK might someday do setup on the main queue doesn't hold any water for me because that is just not realistic.

Reply:
We shouldn't assume which thread this function will be called from, so I agree we should dispatch async to main here.

> make the completion callback on the main queue (not the original queue)

This is a problem for any adapter. We call the partner SDK's init, which then calls our completion, but we can't guarantee that completion is called on the original thread. I just tested this with `GADMobileAds.sharedInstance().start`, and sure enough, the completion is called on the main thread.

This shouldn't be a problem though, since the SDK will dispatch back to its background thread when it returns:

```
adapter.setUp(with: configuration) { [weak self, adapter] error in
    guard let self else { return }
        self.taskDispatcher.async(on: .background) {
           ....
```